README.md
# NAME

Carmel - CPAN Artifact Repository Manager

# SYNOPSIS

    # Run with a directory with cpanfile
    carmel install

    # list all the modules to be loaded
    carmel list

    # list all the modules in a tree
    carmel tree

    # show a location where a module is installed
    carmel show Plack

    # update Plack to the latest
    carmel update Plack

    # update all the modules in the snapshot
    carmel update

    # pin modules tp specific versions
    carmel update DBI@1.633 Plack@1.0000

    # show diffs for cpanfile.snapshot in a nice way
    carmel diff

    # Runs your perl script with modules from artifacts
    carmel exec perl ...

    # Requires all your modules in cpanfile in one shot
    carmel exec perl -e 'use Carmel::Preload;'

    # Roll out the currently selected modules into ./local
    carmel rollout

    # package modules tarballs and index into ./vendor/cache
    carmel package

    # use Carmel packages inside a script (without carmel exec)
    perl -e 'use Carmel::Setup; ...'

    # prints export PATH=... etc for shell scripting
    carmel export

    # find a module in a repository
    carmel find DBI

    # find a module matching the version query
    carmel find Plack ">= 1.0000, < 1.1000"

# DESCRIPTION

Carmel is yet another CPAN module manager.

Unlike traditional CPAN module installer, Carmel keeps the build of
your dependencies in a central repository, then select the library
paths to include upon runtime in development.

Carmel also allows you to rollout all the files in a traditional perl INC
directory structure, which is useful to use in a production environment, such as
containers.

# WORKFLOW

## Development

During the development, run `carmel install` when you setup a new environment,
and any time you make changes to `cpanfile`. This will update your build
artifacts, and saves the changes to `cpanfile.snapshot`. Commit the snapshot
file in version control system so that you can reproduce the exact same versions
across machines.

`carmel exec` makes it easy to run your application using the versions in
`cpanfile` and `cpanfile.snapshot` dynamically.

    # On your development environment
    > cat cpanfile
    requires 'Plack', '0.9980';
    requires 'Starman', '0.2000';

    > carmel install
    > echo /.carmel >> .gitignore
    > git add cpanfile cpanfile.snapshot .gitignore
    > git commit -m "add Plack and Starman"

    # On a new setup, or another developer's machine
    > git pull
    > carmel install
    > carmel exec starman -p 8080 myapp.psgi

    # Add a new dependency
    > echo "requires 'Try::Tiny';" >> cpanfile
    > carmel install
    > git commit -am 'Add Try::Tiny'

    # Update Plack to the latest
    > carmel update Plack

## Production Deployments

Carmel allows you to manage all the dependencies the same way across development
environments and production environments. However, there might be cases where
you want to avoid running your application with `carmel exec` in production, to
avoid the overhead with large number of include paths, or to avoid installing
Carmel in the production hosts. Carmel provides two easy ways to avoid depending
on Carmel on the deploy target environments.

### carmel rollout

`carmel rollout` rolls out the build artifacts into a regular perl5 library
path in `local`. Once the rollout is complete, you can include the path just
like a regular [local::lib](https://metacpan.org/pod/local%3A%3Alib) directory.

    # Production environment: Roll out to ./local
    > carmel rollout
    > perl -Ilocal/lib/perl5 local/bin/starman -p 8080 myapp.psgi

You can run `carmel rollout` in a CI system to create the `local` directory
next to your application code for a linux package (e.g. deb package), or Docker
containers.

### carmel package

`carmel package` (similar to `carton bundle`) creates a directory with
tarballs and CPAN-style package index files, which you can pass to [cpanm](https://metacpan.org/pod/cpanm) on a
target machine. This way, you only need `cpanm`, which is available as a
self-contained single executable, to bootstrap the installation on a host with a
stock perl.

    # Vendor all the packages to vendor/cache
    > carmel package
    > git add vendor/cache
    > git commit -m 'Vendor all the tarballs'

    # Remote environment (CI etc.)
    > git clone https://.../myapp.git && cd myapp

    # Install the modules to ./local (like carmel rollout)
    > cpanm -L ./local --from file://$PWD/vendor/cache -nq --installdeps .

# HOW IT WORKS

Carmel will keep the build directory (artifacts) after a cpanm
installation in a repository, which defaults to `$HOME/.carmel/{version}-{archname}/builds`,
and your directory structure would look like:

    $HOME/.carmel/5.20.1-darwin-2level/builds
      Plack-1.0033/
        blib/
          arch/
          lib/
      URI-1.64/
        blib/
          arch/
          lib/
      URI-1.63/
        blib/
          arch/
          lib/

Carmel scans this directory and creates the mapping of which version
of any package belongs to which build directory.

Given the list of modules and requirements from `cpanfile`, `carmel install`
computes which versions satisfy the requirements best, and if there isn't,
installs the modules from CPAN to put it to the artifact repository. The
computed mappings are preserved as a snapshot in `cpanfile.snapshot`.

Once the snapshot is created, each following `carmel` command runs uses both
`cpanfile` and `cpanfile.snapshot` to determine the best versions to satisfy
the requirements. When you update `cpanfile` to bump a version or add a new
module, `carmel` will install the new dependencies and update the snapshot
accordingly.

`carmel exec` command, like `install` command, lists the build directories and
`.pm` files you need from the repository, and then prepend the mappings of
these files in the `@INC` hook. This is a handy way to run a perl program using
the dependencies pinned by Carmel, without changing any include path.

`carmel update` command allows you to selectively update a dependency while
preserving other dependencies in the snapshot. `carmel update Plack` for
example pulls the latest version of Plack from CPAN (and its dependencies, if it
needs a newer version than pinned in the snapshot), and updates the snapshot
properly. Running `carmel update` without any arguments would update all the
modules in `cpanfile`, including its dependencies.

On a production environment, you might want to use the `carmel rollout`
command, which saves all the files included in the `cpanfile`, pinned with
`cpanfile.snapshot`, to the `local` directory. This directory can be included
like a regular perl's library path, with `PERL5LIB=/path/to/local/lib/perl5`,
or with `use lib`, and you don't need to use `carmel` command in production
this way.

## SNAPSHOT SUPPORT

As of v0.1.29, Carmel supports saving and loading snapshot file in
`cpanfile.snapshot`, in a compatible format with [Carton](https://metacpan.org/pod/Carton). Versions
saved in the snapshot file will be preserved across multiple runs of
Carmel across machines, so that versions frozen in one environment can
be committed to a source code repository, and can be reproduced in
another box, so long as the perl version and architecture is the same.

# CAVEATS / KNOWN ISSUES

- If you run multiple instances of `carmel`, or hit Ctrl-C to interrupt the cpanm
install session, Carmel might get into a state where some modules have been
installed properly, while some modules in the dependency chain are
missing. Carmel checks if there's another process running simultaneously using a
lock file to prevent this problem, but make sure you let it finish the
installation to get the full builds properly.
- There're certain dependencies that get missed during the initial `carmel
install`, and you'll see the error message "Can't find an artifact for
Foo".

    Please report it to the issue tracker if you can reliably reproduce this type of
    errors. [https://github.com/miyagawa/Carmel/issues/74](https://github.com/miyagawa/Carmel/issues/74) has a list of known
    modules that could cause problems like this.

- In some situation, you might encounter conflicts in version resolutions, between
what's pinned in the snapshot and a new version that's needed when you introduce
a new module.

    For example, suppose you have:

        # cpanfile
        requires 'Foo';
        requires 'Bar'; # which requires Foo >= 1.001

    Without a snapshot file, Carmel has no trouble resolving the correct versions
    for this combination. But if you have:

        # cpanfile.snapshot
        Foo-1.000

    The first time you run `carmel install`, Carmel will try to install Foo-1.000,
    because that's the version pinned in the snapshot, while trying to pull the
    module Bar, which would conflict with that version of Foo.

    This can happen 50% of the time, because if cpanm (called internally by Carmel)
    installs Bar first, then the resolution is done correctly and the version in the
    snapshot would be skipped, and the snapshot will be updated accordingly. This is
    due to perl's hash randomization after Perl 5.18.

    To avoid this, you're recommended to run `carmel install` **before making any
    changes to cpanfile**. That will put the build caches to satisfy what's in
    cpanfile and the snapshot. After that, adding a new dependency will likely reuse
    what's in the build cache, while adding a new dependency can update the
    transient dependency (for Foo) without having conflicts.

    If you encounter conflicts like this, you can work around it by:

    - Run `carmel update Foo` to pull the latest version of Foo from CPAN, ignoring what's in the snapshot.
    - Update `cpanfile` to explicitly update the version requirement for `Foo`.

- Carmel doesn't support Taint mode (`-T`). You'll see an error message
`Insecure dependency in require while running with -T switch`.

# COMPARISONS WITH SIMILAR TOOLS

## Carton

Carmel shares the same goal with Carton, where you can manage your dependencies
by declaring them in `cpanfile`, and pinning them in `cpanfile.snapshot`. Most
of the commands work the same way, so Carmel can most effectively be a drop-in
replacement for Carton, if you're currently using it.

Here's a few key differences between Carmel and Carton:

- Carton _does not_ manage what's currently being installed in `local`
directory. It just runs `cpanm` command with `-L local`, with a hope that
nothing has changed the directory except Carton, and whatever is in the
directory won't conflict with the snapshot file. This can easily conflict when
`cpanfile.snapshot` is updated by multiple developers or when you continuously
update the dependencies across multiple machines.

    Carmel manages all the dependencies for your project in the Carmel repository
    under `$HOME/.carmel`, and nothing is installed under your project directory on
    development. The `local` directory is only created when you request it via
    `carmel rollout` command, and it's safe to run multiple times. Running `carmel
    install` after pulling the changes to the snapshot file will always install the
    correct dependencies from the snapshot file, as compared to Carton, which
    doesn't honor the snapshot on a regular install command, if whatever version in
    `local` already satisfies the version in `cpanfile`.

- Carton has no easy way to undo a change once you update a version of a module in
`local`, because which version is actually selected is only preserved as a file
inside the directory, that's not managed by Carton. To undo a change you have to
remove the entire `local` directory to start over.

    Carmel preserves this information to the `cpanfile.snapshot` file, and every
    invocation of Carmel resolves the dependencies declared in `cpanfile` and
    pinned in `cpanfile.snapshot` dynamically, to create a stable dependency tree,
    without relying on anything in a directory under your project other than the
    snapshot file. Undoing the change in `cpanfile.snapshot` file immediately
    reverts the change.

## cpm

[App::cpm](https://metacpan.org/pod/App%3A%3Acpm) is an excellent standalone CPAN installer.

- Like [Carton](https://metacpan.org/pod/Carton), cpm installs the dependencies declared in `cpanfile` to
`local`. Carmel installs them into a build cache, and doesn't use `local`
directory for state management. You can run `carmel rollout` to copy the
dependencies to `local` directory.
- cpm installs the modules in parallel, which makes the installation very
fast. Like Carmel, cpm also manages its build artifacts cache, so a module that
has previously been installed would be very fast to install, since there's no build
process.
- Unlike Carton and Carmel, cpm doesn't have the ability to manage
`cpanfile.snapshot` file on its own. It can read the snapshot however, so it's
possible to use Carmel in a development environment, and then use `cpm
install` instead of `carmel install` and `carmel rollout`, if all you need is
to build out a perl5 library path out of `cpanfile` and `cpanfile.snapshot` in
the source code repository.

# COMMUNITY

- [https://github.com/miyagawa/Carmel](https://github.com/miyagawa/Carmel)

    Code repository, Wiki and Issue Tracker

# AUTHOR

Tatsuhiko Miyagawa <miyagawa@bulknews.net>

# COPYRIGHT

Copyright 2015- Tatsuhiko Miyagawa

# LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# SEE ALSO

[App::cpanminus](https://metacpan.org/pod/App%3A%3Acpanminus) [Carton](https://metacpan.org/pod/Carton)


Changes
Revision history for Carmel

{{$NEXT}}
        - changed dependency on JSON to JSON::MaybeXS

v0.9.2  2022-05-27 00:31:06 PDT
    [IMPROVEMENTS]
        - carmel update is now much faster, especially when a custom mirror is used (#58)
        - uses git diff internally for git coloring rather than janky regexes (#87)
        - internal refactoring

v0.9.1  2022-05-20 14:55:34 PDT
    [IMPROVEMENTS]
        - Added a file-based lock during install/update run to avoid corrupted artifacts (#41, #79)
        - Improved error message for running under -T (#76)

v0.9.0  2022-05-14 15:59:45 PDT
    [NEW FEATURES]
        - Added carmel reinstall to rebuild build artifacts for all the dependencies (#59, #68)
        - Added carmel info as an alias for show
        - Added carmel look to open a new shell inside the build artifact directory, like cpanm --look
        - Added carmel run, like carmel exec but uses system() instead of exec() (#20)
    [BUG FIXES]
        - Ignore core modules that are not dual life e.g. strict.pm (#69)
    [IMPROVEMENTS]
        - carmel update won't downgrade if you have a higher version than what's on CPAN (#66)
        - carmel update progress bar will show the actual number of dists, rather than percentage
        - Uses https mirror (https://cpan.metacpan.org/) by default (#42)
        - Remove existing artifacts before importing, for clean install (#70)
        - carmel inject now works without cpanfile, outside a project directory (#73)
        - Documentation updates

v0.1.56  2022-05-11 12:40:21 PDT
        - carmel install without the snapshot will install the latest versions (#53)

v0.1.55  2022-05-11 00:20:03 PDT
        - carmel update now updates core modules specified in cpanfile (#48)
        - Fix a bug where carmel update fails with division by zero when there's no dist in snapshot

v0.1.54  2022-05-10 21:54:59 PDT
        - Add a simple progress bar to carmel update (#56)

v0.1.53  2022-05-10 02:16:50 PDT
        - Added new command carmel diff to nicely format cpanfile.snapshot diff (#54)

v0.1.52  2022-05-09 12:51:21 PDT
        - Fix a bug where carmel update fails when custom mirror is specified in cpanfile (#52)

v0.1.51  2022-05-09 09:59:21 PDT
        - Unified carmel update/pin and deprecated the pin command.
        - carmel update now runs much faster if you already have the build artifacts for the new version.
        - Improved the error message when carmel update conflicts with what's in cpanfile
        - Work around a corner case where carmel update fails with bad NAME in MYMETA e.g. LWP (#51)

v0.1.50  2022-05-09 00:23:17 PDT
        - Fixed a bug where core modules in the snapshot can be ignored in the fist carmel install (#47)
        - Carmel now automatically upgrades core modules specified in the snapshot if the version is
          lower than what's in core, and cpanfile doesn't have any specific requirement. (#49)
        - Speed up carmel install when all the required versions are in the artifacts

v0.1.49  2022-05-08 11:47:37 PDT
        - Speed up the artifacts search and snapshot generation with lazy-loading and JSON backend

v0.1.48  2022-05-08 02:44:56 PDT
        - Huge performance improvements for carmel update to avoid installation when there's no updates (#44)

v0.1.47  2022-05-07 22:12:42 PDT
        - Fix carmel package to not fetch from CPAN when cache is empty
        - Internal refactoring of the code base

v0.1.46  2022-05-07 14:10:39 PDT
        - New command: carmel pin to pin a module version in the snapshot without pinning it in cpanfile (#39)
        - Improve error message when a conflict is found
        - Fix a bug where a distribution can appear multiple times when downgrading a distribution (#38)
        - Make artifacts search always strict after install is done
        - Improved documentation

v0.1.45  2022-05-07 01:36:08 PDT
        - Cleanup perl5 temporary library path to avoid missing artifacts when previous
          installation was cancelled
        - Add sort to stablize carmel env/export command outputs

v0.1.44  2022-05-06 20:39:25 PDT
        - Fix a bug where Module::Build::Tiny is not installed on the first run (#34)

v0.1.43  2022-05-06 19:17:34 PDT
        - Dropped support for perl 5.12
        - Removed the experimental, undocumented binstubs command
        - Fixed so carmel command exits with non-zero upon failures
        - Fixed `carmel show Module` to work with sub-modules in a distribution

v0.1.42  2022-05-06 17:02:09 PDT
        - Documentation updates

v0.1.41  2022-05-06 14:35:01 PDT
        - Add support for carmel update Module

v0.1.40  2022-05-06 01:51:17 PDT
        - Remove the unnecessary pinning of versions restored from the snapshot (#33)

v0.1.39  2022-02-12 21:20:22 PST
        - updated docs

v0.1.38  2019-01-10 13:38:17 PST
        - Include configure/runtime dependencies for artifacts to include in cpanfile.snapshot

v0.1.37  2018-04-25 02:43:30 PDT
        - Update Menlo dependency

v0.1.36  2018-04-20 14:54:38 CEST
        - Fix the usage of Menlo::CLI::Compat to follow the change in Menlo

v0.1.35  2018-02-10 01:14:53 PST
        - Fix Carmel::Preload failing to load a module when cpanfile contains specific version
          requirements #31

v0.1.34  2016-09-25 15:24:53 PDT
        - Use cpan.metacpan.org as a default fast/backpan mirror
        - Fix SiteINC to run nested carmel exec #28

v0.1.33  2016-05-25 22:39:03 PDT
        - Same as v0.1.32

v0.1.32  2016-05-24 10:37:10 PDT
        - Allow falling back to backpan when restoring from snapshot

v0.1.31  2016-05-24 01:37:11 PDT
        - Experimental support for mirror in cpanfile
        - Support restoring exact snapshot versions via mirror-index for Menlo

v0.1.30  2016-05-23 13:32:48 PDT
        - Store configure/build requirements in snapshot as well for a compatibility with Carton

v0.1.29  2016-05-22 13:02:57 PDT
        - Fixes various corner cases to preserve snapshot versions

lib/Carmel.pm
package Carmel;
use strict;
use 5.012000;
use version; our $VERSION = version->declare('v0.9.2');

1;
__END__

=encoding utf-8

=head1 NAME

Carmel - CPAN Artifact Repository Manager

=head1 SYNOPSIS

  # Run with a directory with cpanfile
  carmel install

  # list all the modules to be loaded
  carmel list

  # list all the modules in a tree
  carmel tree

  # show a location where a module is installed
  carmel show Plack

  # update Plack to the latest
  carmel update Plack

  # update all the modules in the snapshot
  carmel update

  # pin modules tp specific versions
  carmel update DBI@1.633 Plack@1.0000

  # show diffs for cpanfile.snapshot in a nice way
  carmel diff

  # Runs your perl script with modules from artifacts
  carmel exec perl ...

  # Requires all your modules in cpanfile in one shot
  carmel exec perl -e 'use Carmel::Preload;'

  # Roll out the currently selected modules into ./local
  carmel rollout

  # package modules tarballs and index into ./vendor/cache
  carmel package

  # use Carmel packages inside a script (without carmel exec)
  perl -e 'use Carmel::Setup; ...'

  # prints export PATH=... etc for shell scripting
  carmel export

  # find a module in a repository
  carmel find DBI

  # find a module matching the version query
  carmel find Plack ">= 1.0000, < 1.1000"

=head1 DESCRIPTION

Carmel is yet another CPAN module manager.

Unlike traditional CPAN module installer, Carmel keeps the build of
your dependencies in a central repository, then select the library
paths to include upon runtime in development.

Carmel also allows you to rollout all the files in a traditional perl INC
directory structure, which is useful to use in a production environment, such as
containers.

=head1 WORKFLOW

=head2 Development

During the development, run C<carmel install> when you setup a new environment,
and any time you make changes to C<cpanfile>. This will update your build
artifacts, and saves the changes to C<cpanfile.snapshot>. Commit the snapshot
file in version control system so that you can reproduce the exact same versions
across machines.

C<carmel exec> makes it easy to run your application using the versions in
C<cpanfile> and C<cpanfile.snapshot> dynamically.

  # On your development environment
  > cat cpanfile
  requires 'Plack', '0.9980';
  requires 'Starman', '0.2000';

  > carmel install
  > echo /.carmel >> .gitignore
  > git add cpanfile cpanfile.snapshot .gitignore
  > git commit -m "add Plack and Starman"

  # On a new setup, or another developer's machine
  > git pull
  > carmel install
  > carmel exec starman -p 8080 myapp.psgi

  # Add a new dependency
  > echo "requires 'Try::Tiny';" >> cpanfile
  > carmel install
  > git commit -am 'Add Try::Tiny'

  # Update Plack to the latest
  > carmel update Plack

=head2 Production Deployments

Carmel allows you to manage all the dependencies the same way across development
environments and production environments. However, there might be cases where
you want to avoid running your application with C<carmel exec> in production, to
avoid the overhead with large number of include paths, or to avoid installing
Carmel in the production hosts. Carmel provides two easy ways to avoid depending
on Carmel on the deploy target environments.

=head3 carmel rollout

C<carmel rollout> rolls out the build artifacts into a regular perl5 library
path in C<local>. Once the rollout is complete, you can include the path just
like a regular L<local::lib> directory.

  # Production environment: Roll out to ./local
  > carmel rollout
  > perl -Ilocal/lib/perl5 local/bin/starman -p 8080 myapp.psgi

You can run C<carmel rollout> in a CI system to create the C<local> directory
next to your application code for a linux package (e.g. deb package), or Docker
containers.

=head3 carmel package

C<carmel package> (similar to C<carton bundle>) creates a directory with
tarballs and CPAN-style package index files, which you can pass to L<cpanm> on a
target machine. This way, you only need C<cpanm>, which is available as a
self-contained single executable, to bootstrap the installation on a host with a
stock perl.

  # Vendor all the packages to vendor/cache
  > carmel package
  > git add vendor/cache
  > git commit -m 'Vendor all the tarballs'

  # Remote environment (CI etc.)
  > git clone https://.../myapp.git && cd myapp

  # Install the modules to ./local (like carmel rollout)
  > cpanm -L ./local --from file://$PWD/vendor/cache -nq --installdeps .

=head1 HOW IT WORKS

Carmel will keep the build directory (artifacts) after a cpanm
installation in a repository, which defaults to C<$HOME/.carmel/{version}-{archname}/builds>,
and your directory structure would look like:

  $HOME/.carmel/5.20.1-darwin-2level/builds
    Plack-1.0033/
      blib/
        arch/
        lib/
    URI-1.64/
      blib/
        arch/
        lib/
    URI-1.63/
      blib/
        arch/
        lib/

Carmel scans this directory and creates the mapping of which version
of any package belongs to which build directory.

Given the list of modules and requirements from C<cpanfile>, C<carmel install>
computes which versions satisfy the requirements best, and if there isn't,
installs the modules from CPAN to put it to the artifact repository. The
computed mappings are preserved as a snapshot in C<cpanfile.snapshot>.

Once the snapshot is created, each following C<carmel> command runs uses both
C<cpanfile> and C<cpanfile.snapshot> to determine the best versions to satisfy
the requirements. When you update C<cpanfile> to bump a version or add a new
module, C<carmel> will install the new dependencies and update the snapshot
accordingly.

C<carmel exec> command, like C<install> command, lists the build directories and
C<.pm> files you need from the repository, and then prepend the mappings of
these files in the C<@INC> hook. This is a handy way to run a perl program using
the dependencies pinned by Carmel, without changing any include path.

C<carmel update> command allows you to selectively update a dependency while
preserving other dependencies in the snapshot. C<carmel update Plack> for
example pulls the latest version of Plack from CPAN (and its dependencies, if it
needs a newer version than pinned in the snapshot), and updates the snapshot
properly. Running C<carmel update> without any arguments would update all the
modules in C<cpanfile>, including its dependencies.

On a production environment, you might want to use the C<carmel rollout>
command, which saves all the files included in the C<cpanfile>, pinned with
C<cpanfile.snapshot>, to the C<local> directory. This directory can be included
like a regular perl's library path, with C<PERL5LIB=/path/to/local/lib/perl5>,
or with C<use lib>, and you don't need to use C<carmel> command in production
this way.

=head2 SNAPSHOT SUPPORT

As of v0.1.29, Carmel supports saving and loading snapshot file in
C<cpanfile.snapshot>, in a compatible format with L<Carton>. Versions
saved in the snapshot file will be preserved across multiple runs of
Carmel across machines, so that versions frozen in one environment can
be committed to a source code repository, and can be reproduced in
another box, so long as the perl version and architecture is the same.

=head1 CAVEATS / KNOWN ISSUES

=over 4

=item *

If you run multiple instances of C<carmel>, or hit Ctrl-C to interrupt the cpanm
install session, Carmel might get into a state where some modules have been
installed properly, while some modules in the dependency chain are
missing. Carmel checks if there's another process running simultaneously using a
lock file to prevent this problem, but make sure you let it finish the
installation to get the full builds properly.

=item *

There're certain dependencies that get missed during the initial C<carmel
install>, and you'll see the error message "Can't find an artifact for
Foo".

Please report it to the issue tracker if you can reliably reproduce this type of
errors. L<https://github.com/miyagawa/Carmel/issues/74> has a list of known
modules that could cause problems like this.

=item *

In some situation, you might encounter conflicts in version resolutions, between
what's pinned in the snapshot and a new version that's needed when you introduce
a new module.

For example, suppose you have:

  # cpanfile
  requires 'Foo';
  requires 'Bar'; # which requires Foo >= 1.001

Without a snapshot file, Carmel has no trouble resolving the correct versions
for this combination. But if you have:

  # cpanfile.snapshot
  Foo-1.000

The first time you run C<carmel install>, Carmel will try to install Foo-1.000,
because that's the version pinned in the snapshot, while trying to pull the
module Bar, which would conflict with that version of Foo.

This can happen 50% of the time, because if cpanm (called internally by Carmel)
installs Bar first, then the resolution is done correctly and the version in the
snapshot would be skipped, and the snapshot will be updated accordingly. This is
due to perl's hash randomization after Perl 5.18.

To avoid this, you're recommended to run C<carmel install> B<before making any
changes to cpanfile>. That will put the build caches to satisfy what's in
cpanfile and the snapshot. After that, adding a new dependency will likely reuse
what's in the build cache, while adding a new dependency can update the
transient dependency (for Foo) without having conflicts.

If you encounter conflicts like this, you can work around it by:

=over 8

=item *

Run C<carmel update Foo> to pull the latest version of Foo from CPAN, ignoring what's in the snapshot.

=item *

Update C<cpanfile> to explicitly update the version requirement for C<Foo>.

=back

=item *

Carmel doesn't support Taint mode (C<-T>). You'll see an error message
C<Insecure dependency in require while running with -T switch>.

=back

=head1 COMPARISONS WITH SIMILAR TOOLS

=head2 Carton

Carmel shares the same goal with Carton, where you can manage your dependencies
by declaring them in C<cpanfile>, and pinning them in C<cpanfile.snapshot>. Most
of the commands work the same way, so Carmel can most effectively be a drop-in
replacement for Carton, if you're currently using it.

Here's a few key differences between Carmel and Carton:

=over 4

=item *

Carton I<does not> manage what's currently being installed in C<local>
directory. It just runs C<cpanm> command with C<-L local>, with a hope that
nothing has changed the directory except Carton, and whatever is in the
directory won't conflict with the snapshot file. This can easily conflict when
C<cpanfile.snapshot> is updated by multiple developers or when you continuously
update the dependencies across multiple machines.

Carmel manages all the dependencies for your project in the Carmel repository
under C<$HOME/.carmel>, and nothing is installed under your project directory on
development. The C<local> directory is only created when you request it via
C<carmel rollout> command, and it's safe to run multiple times. Running C<carmel
install> after pulling the changes to the snapshot file will always install the
correct dependencies from the snapshot file, as compared to Carton, which
doesn't honor the snapshot on a regular install command, if whatever version in
C<local> already satisfies the version in C<cpanfile>.

=item *

Carton has no easy way to undo a change once you update a version of a module in
C<local>, because which version is actually selected is only preserved as a file
inside the directory, that's not managed by Carton. To undo a change you have to
remove the entire C<local> directory to start over.

Carmel preserves this information to the C<cpanfile.snapshot> file, and every
invocation of Carmel resolves the dependencies declared in C<cpanfile> and
pinned in C<cpanfile.snapshot> dynamically, to create a stable dependency tree,
without relying on anything in a directory under your project other than the
snapshot file. Undoing the change in C<cpanfile.snapshot> file immediately
reverts the change.

=back

=head2 cpm

L<App::cpm> is an excellent standalone CPAN installer.

=over 4

=item *

Like L<Carton>, cpm installs the dependencies declared in C<cpanfile> to
C<local>. Carmel installs them into a build cache, and doesn't use C<local>
directory for state management. You can run C<carmel rollout> to copy the
dependencies to C<local> directory.

=item *

cpm installs the modules in parallel, which makes the installation very
fast. Like Carmel, cpm also manages its build artifacts cache, so a module that
has previously been installed would be very fast to install, since there's no build
process.

=item *

Unlike Carton and Carmel, cpm doesn't have the ability to manage
C<cpanfile.snapshot> file on its own. It can read the snapshot however, so it's
possible to use Carmel in a development environment, and then use C<cpm
install> instead of C<carmel install> and C<carmel rollout>, if all you need is
to build out a perl5 library path out of C<cpanfile> and C<cpanfile.snapshot> in
the source code repository.

=back

=head1 COMMUNITY

=over 4

=item L<https://github.com/miyagawa/Carmel>

Code repository, Wiki and Issue Tracker

=back

=head1 AUTHOR

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt>

=head1 COPYRIGHT

Copyright 2015- Tatsuhiko Miyagawa

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<App::cpanminus> L<Carton>

=cut
